File: src/app/schemas/lms.py
from datetime import datetime

from pydantic import BaseModel, EmailStr, Field


class UserProfile(BaseModel):
    id: str
    username: str
    full_name: str
    role: str
    status: str
    email: EmailStr
    group_name: str = ""
    department: str = ""


class DemoAccount(BaseModel):
    username: str
    password: str
    full_name: str
    role: str
    status: str


class AuthLoginRequest(BaseModel):
    username: str
    password: str


class AuthSession(BaseModel):
    access_token: str
    token_type: str = "bearer"
    expires_at: datetime
    user: UserProfile


class VisitLeadIn(BaseModel):
    name: str
    email: EmailStr
    phone: str
    message: str


class VisitLeadOut(BaseModel):
    id: str
    created_at: datetime


class ExternalUserLink(BaseModel):
    id: str
    external_system: str
    external_user_id: str
    external_username: str
    linked_at: datetime


class ScheduleSlot(BaseModel):
    id: str
    course_id: str
    subject: str
    room: str
    group_name: str
    day_of_week: int
    starts_at: str
    ends_at: str
    teacher_name: str


class NotificationItem(BaseModel):
    id: str
    kind: str
    title: str
    body: str
    entity_type: str
    entity_id: str
    created_at: datetime
    is_read: bool = False


class CourseSummary(BaseModel):
    id: str
    title: str
    description: str
    teacher_name: str
    audience: str
    status: str
    progress: int = 0


class CourseModule(BaseModel):
    id: str
    title: str
    content: str
    position: int
    estimated_minutes: int
    completed: bool = False
    completed_at: datetime | None = None


class TestSummary(BaseModel):
    id: str
    course_id: str
    title: str
    description: str
    duration_minutes: int
    max_attempts: int
    passing_score: int
    latest_status: str = "new"
    latest_score: int | None = None


class CourseDetail(BaseModel):
    course: CourseSummary
    modules: list[CourseModule]
    tests: list[TestSummary]


class TestQuestion(BaseModel):
    id: str
    prompt: str
    options: list[str]
    question_type: str
    position: int
    points: int


class TestAttemptStart(BaseModel):
    attempt_id: str
    test: TestSummary
    questions: list[TestQuestion]
    started_at: datetime


class TestAnswerSubmission(BaseModel):
    question_id: str
    selected_option_ids: list[int] = Field(default_factory=list)


class TestAttemptSubmission(BaseModel):
    answers: list[TestAnswerSubmission]


class TestAttemptResult(BaseModel):
    attempt_id: str
    status: str
    score: int
    passing_score: int
    earned_points: int
    total_points: int
    submitted_at: datetime


class ReviewQueueItem(BaseModel):
    attempt_id: str
    student_name: str
    course_title: str
    test_title: str
    score: int
    submitted_at: datetime
    teacher_name: str


class DashboardPayload(BaseModel):
    user: UserProfile
    courses: list[CourseSummary]
    schedule: list[ScheduleSlot]
    notifications: list[NotificationItem]
    external_links: list[ExternalUserLink]
    review_queue: list[ReviewQueueItem]
    demo_accounts: list[DemoAccount]

File: src/types/lms.ts
export type UserRole = 'admin' | 'student' | 'teacher'

export type AccountStatus = 'approved' | 'pending' | 'blocked'

export type PageId =
  | 'home'
  | 'courses'
  | 'tests'
  | 'schedule'
  | 'reviews'
  | 'notifications'
  | 'links'
  | 'profile'

export type NavItem = {
  id: PageId
  label: string
}

export type UserProfile = {
  id: string
  username: string
  full_name: string
  role: UserRole
  status: AccountStatus
  email: string
  group_name: string
  department: string
}

export type DemoAccount = {
  username: string
  password: string
  full_name: string
  role: UserRole
  status: AccountStatus
}

export type AuthSession = {
  access_token: string
  token_type: string
  expires_at: string
  user: UserProfile
}

export type CourseSummary = {
  id: string
  title: string
  description: string
  teacher_name: string
  audience: string
  status: string
  progress: number
}

export type CourseModule = {
  id: string
  title: string
  content: string
  position: number
  estimated_minutes: number
  completed: boolean
  completed_at: string | null
}

export type TestSummary = {
  id: string
  course_id: string
  title: string
  description: string
  duration_minutes: number
  max_attempts: number
  passing_score: number
  latest_status: 'new' | 'in_progress' | 'submitted' | 'failed' | 'passed'
  latest_score: number | null
}

export type CourseDetail = {
  course: CourseSummary
  modules: CourseModule[]
  tests: TestSummary[]
}

export type TestQuestion = {
  id: string
  prompt: string
  options: string[]
  question_type: 'single' | 'multiple'
  position: number
  points: number
}

export type TestAttemptStart = {
  attempt_id: string
  started_at: string
  test: TestSummary
  questions: TestQuestion[]
}

export type TestAnswerSubmission = {
  question_id: string
  selected_option_ids: number[]
}

export type TestAttemptResult = {
  attempt_id: string
  status: 'passed' | 'failed'
  score: number
  passing_score: number
  earned_points: number
  total_points: number
  submitted_at: string
}

export type ScheduleSlot = {
  id: string
  course_id: string
  subject: string
  room: string
  group_name: string
  day_of_week: number
  starts_at: string
  ends_at: string
  teacher_name: string
}

export type NotificationItem = {
  id: string
  kind: string
  title: string
  body: string
  entity_type: string
  entity_id: string
  created_at: string
  is_read: boolean
}

export type ExternalUserLink = {
  id: string
  external_system: string
  external_user_id: string
  external_username: string
  linked_at: string
}

export type ReviewQueueItem = {
  attempt_id: string
  student_name: string
  course_title: string
  test_title: string
  score: number
  submitted_at: string
  teacher_name: string
}

export type DashboardPayload = {
  user: UserProfile
  courses: CourseSummary[]
  schedule: ScheduleSlot[]
  notifications: NotificationItem[]
  external_links: ExternalUserLink[]
  review_queue: ReviewQueueItem[]
  demo_accounts: DemoAccount[]
}
